rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java

```diff
@@ -152,224 +137,213 @@ public void processBlockHeaders(@Nonnull final MessageSender sender, @Nonnull fi
+     * @param requestId the id of the request.
+     * @param startNumber the starting block's hash to get the skeleton.
+     */
+    @Override
+    public void processSkeletonRequest(@Nonnull final MessageChannel sender, long requestId, long startNumber) {
+        logger.trace("Processing block hash request {} {} {} from {}", requestId, startNumber, sender.getPeerNodeID().toString());
+        // TODO(mc) move this to configuration
+        int skeletonStep = 192;
+        Block blockStart = this.getBlockFromBlockchainStore(startNumber);
+
+        // If we don't have a block with the requested number, we ignore the message
+        if (blockStart == null) {
+            // Don't waste time sending an empty response.
+            return;
+        }
 
-        for (long n = peerBestBlockNumber; n <= bestBlockNumber && n < peerBestBlockNumber + this.blocksForPeers; n++) {
-            logger.trace("Trying to send block {}", n);
-            
-            final Block b = this.blockchain.getBlockByNumber(n);
+        // We always include the skeleton block immediately before blockStart, even if it's Genesis
+        long skeletonStartHeight = (blockStart.getNumber() / skeletonStep) * skeletonStep;
+        List<BlockIdentifier> blockIdentifiers = new ArrayList<>();
+        for (long skeletonNumber = skeletonStartHeight; skeletonNumber < this.getBestBlockNumber(); skeletonNumber += skeletonStep) {
+            byte[] skeletonHash = getSkeletonHash(skeletonNumber);
+            blockIdentifiers.add(new BlockIdentifier(skeletonHash, skeletonNumber));
+        }
 
-            if (b == null)
-                continue;
+        // We always include the best block as part of the Skeleton response
+        blockIdentifiers.add(new BlockIdentifier(this.getBestBlockHash(), this.getBestBlockNumber()));
+        SkeletonResponseMessage responseMessage = new SkeletonResponseMessage(requestId, blockIdentifiers);
 
-            nodeInformation.addBlockToNode(new ByteArrayWrapper(b.getHash()), sender.getNodeID());
-            logger.trace("Sending block {} {}", b.getNumber(), b.getShortHash());
-            sender.sendMessage(new BlockMessage(b));
-        }
+        sender.sendMessage(responseMessage);
     }
 
     /**
-     * processGetBlock sends a requested block to a peer if the block is available.
      *
-     * @param sender the sender of the GetBlock message.
-     * @param hash   the requested block's hash.
+     * @param skeletonBlockNumber a block number that belongs to the skeleton
+     * @return the proper hash for the block
      */
-    @Override
-    public void processGetBlock(@Nonnull final MessageSender sender, @Nullable final byte[] hash) {
-        logger.trace("Processing get block " + Hex.toHexString(hash).substring(0, 10) + " from " + sender.getNodeID().toString());
-        final Block block = this.getBlock(hash);
-
-        if (block == null) {
-            return;
+    private byte[] getSkeletonHash(long skeletonBlockNumber) {
```

Comment:
Is this method `fork aware`??

Reply:
Improved implementation, now it doesn't cache a hash if it too close to the end of the blockchain